generative_data_prep/processors/article_tokenizer.py

```diff
@@ -141,6 +148,200 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
+
+        tokenized_articles = []
+        tokens = []
+        category_id = -1
+        orginal_prompt_completion = None
+        prompt_index = 0
+
+        new_prompt = ""
+        new_completion = ""
+
+        """
+        Automatic conversion back to [prompt, completion] format, especially if prompt_loss_weight
+        is set during the training phase.
+        1. For standard prompt and completion tokens, the existing pipeline is retained.
+        2. Tokens introduced by the chat template are classified as completion tokens.
+        """
+        while split_chat:
+            token = split_chat.pop(0)
+            # Check the start to see if it is a prompt or a completion.
+            if start_instr_of_chat and token == start_instr_of_chat:
+                new_prompt = start_instr_of_chat
+
+            if token == DEFAULT_PROMPT_PLACEHOLDER:
+                if not prompts:
+                    err_msg = f"Prompts list is empty, but {DEFAULT_PROMPT_PLACEHOLDER} found in split_chat"
+                    raise ValueError(err_msg)
+                prompt = prompts.pop(0)
+
+                orginal_prompt_completion = jsonl[prompt_index]
+                if prompt != orginal_prompt_completion[self.prompt_keyword]:
+                    err_msg = f"Prompt mismatch, {prompt} != {orginal_prompt_completion[self.prompt_keyword]}"
+                    raise ValueError(err_msg)
+
+                category_id = self.get_category_id(orginal_prompt_completion)
+                # tokens += self.tokenize(completion=None, prompt=prompt, category_id=category_id)
+                prompt_index += 1
+
+                new_prompt += prompt
+                while split_chat[0] != DEFAULT_COMPLETION_PLACEHOLDER:
+                    new_prompt += split_chat.pop(0)
+
+            elif token == DEFAULT_COMPLETION_PLACEHOLDER:
+                if not completions:
+                    err_msg = f"Completions list is empty, but {DEFAULT_COMPLETION_PLACEHOLDER} found in split_chat"
+                    raise ValueError(err_msg)
+                completion = completions.pop(0)
+
+                if (orginal_prompt_completion is not None) and completion != orginal_prompt_completion[
+                    self.completion_keyword
+                ]:
+                    err_msg = (
+                        f"Completion mismatch, {completion} != {orginal_prompt_completion[self.completion_keyword]}"
+                    )
+                    raise ValueError(err_msg)
+
+                new_completion = completion
+                while split_chat and (
+                    split_chat[0] != DEFAULT_PROMPT_PLACEHOLDER
+                    or (start_instr_of_chat and split_chat[0] != start_instr_of_chat)
+                ):
+                    new_completion += split_chat.pop(0)
+
+                # remove the duplicated eos tokens from new_completion as we will add it later
+                new_completion = new_completion.replace(self.tokenizer.eos_token, "")
```

Comment:
Will there be a corner case if eos is in the middle of the completion?

Reply:
thanks, good point, I am going to fix it in the latest commit